metadata-ingestion/src/datahub/ingestion/source/sql/snowflake.py

```diff
@@ -214,34 +224,214 @@ def get_identifier(self, *, schema: str, entity: str, **kwargs: Any) -> str:
+        # NOTE: This query captures only the upstream lineage of a view.
+        # For more details see https://docs.snowflake.com/en/sql-reference/account-usage/access_history.html#usage-notes for current limitations on capturing the lineage for views.
+        view_upstream_lineage_query: str = """
+WITH view_upstream_lineage_history AS (
+  SELECT
+    vu.value : "objectName" AS view_name,
+    vu.value : "objectDomain" AS view_domain,
+    vu.value : "columns" AS view_columns,
+    r.value : "objectName" AS upstream_table_name,
+    r.value : "objectDomain" AS upstream_table_domain,
+    r.value : "columns" AS upstream_table_columns,
+    t.query_start_time AS query_start_time
+  FROM
+    (
+      SELECT
+        *
+      FROM
+        snowflake.account_usage.access_history
+    ) t,
+    lateral flatten(input => t.DIRECT_OBJECTS_ACCESSED) vu,
+    lateral flatten(input => t.BASE_OBJECTS_ACCESSED) r
+  WHERE
+    vu.value : "objectId" IS NOT NULL
+    AND r.value : "objectId" IS NOT NULL
+    AND t.query_start_time >= to_timestamp_ltz({start_time_millis}, 3)
+    AND t.query_start_time < to_timestamp_ltz({end_time_millis}, 3)
+)
+SELECT
+  view_name,
+  view_columns,
+  upstream_table_name,
+  upstream_table_domain,
+  upstream_table_columns
+FROM
+  view_upstream_lineage_history
+WHERE
+  view_domain in ('View', 'Materialized view')
+  AND view_name != upstream_table_name
+  QUALIFY ROW_NUMBER() OVER (
+    PARTITION BY view_name,
+    upstream_table_name
+    ORDER BY
+      query_start_time DESC
+  ) = 1
+        """.format(
+            start_time_millis=int(self.config.start_time.timestamp() * 1000),
+            end_time_millis=int(self.config.end_time.timestamp() * 1000),
+        )
+
+        assert self._lineage_map is not None
+
+        try:
+            for db_row in engine.execute(view_upstream_lineage_query):
+                # Process UpstreamTable->View edge.
+
+                view_name: str = db_row[0].lower().replace('"', "")
+                upstream_table: str = db_row[2].lower().replace('"', "")
+                # key is the downstream view name
+                self._lineage_map[view_name].append(
+                    # (<upstream_table_name>, <json_list_of_upstream_table_columns>, <json_list_of_downstream_view_columns>)
+                    (upstream_table, db_row[4], db_row[1])
+                )
+                logger.debug(
+                    f"Table->View: Lineage[{view_name}]:{self._lineage_map[view_name]}, upstream_domain={db_row[3]}"
+                )
+        except Exception as e:
+            logger.warning(
+                f"Extracting the upstream view lineage from Snowflake failed."
+                f"Please check your permissions. Continuing...\nError was {e}."
+            )
+
+    def _populate_view_downstream_lineage(
+        self, engine: sqlalchemy.engine.Engine
+    ) -> None:
+        # NOTE: This query captures both the upstream and downstream table lineage for views.
```

Comment:
if this query captures both upstream & downstream lineage, why do we need the above method?

Reply:
The first query is required for capturing the upstream lineage of the views that don't have any downstream lineage (it captures upstream lineage for views that have downstream lineage as well). The second query is to specifically capture the downstream table lineage of views as well as removing false upstream table to the downstream table edges when there is a view between them(snowflake access history is a bit contrived with views).